[FIXED JENKINS-26964] Handle AbortException right

KostyaSha opened the pull request:
[JENKINS-26964](https://issues.jenkins-ci.org/browse/JENKINS-26964)

KostyaSha committed:
[FIXED JENKINS-26964] Handle AbortException right

jtnord commented on core/src/main/java/hudson/model/AbstractBuild.java:
```diff
@@ -733,7 +734,9 @@ protected final boolean performAllBuildSteps(BuildListener listener, Iterable<?
+            if (!(e instanceof AbortException)){
+                e.printStackTrace(listener.error(msg));
+            }
             LOGGER.log(WARNING, msg, e);
```
why print the stack trace in the log for Aborts?

KostyaSha replied:
Read code again.

jtnord replied:
I'm not talking about the build log (listener) but the Jenkins log Logger.log(..) which is called outside the if() statement

KostyaSha replied:
Hm, i can move it to if.

KostyaSha commented:
Fixed comment, fixup'ed into one commit.

ikedam commented:
How about overriding `AbortException#printStacktrace()`?
* It can print the message without the stacktrace.
* No need to handle `AbortException` everywhere in Jenkins.

KostyaSha commented:
@ikedam :-1:

jtnord commented:
:+1:

oleg-nenashev commented:
@ikedam 
I would print <code>AbortException</code> contents to the log in any case. Probably, for code>AbortException</code> stack-traces it makes sense to change the logging level to INFO or FINE in such case

KostyaSha commented:
@oleg-nenashev to which log? No need to print stacktrace into listener, this is known and handled exception.

orrc commented on test/src/test/java/hudson/model/FreestyleJobPublisherTest.java:
```diff
@@ -0,0 +1,89 @@
+        p.getPublishersList().add(new ArtifactArchiver("result.txt", "", false));
+
+        FreeStyleBuild b = p.scheduleBuild2(0).get();
+        assertEquals("Build must fail, because we used FalsePublisher", b.getResult(), Result.FAILURE);
```
`FalsePublisher` -> `AbortExceptionPublisher`

KostyaSha replied:
Thanks.

orrc commented on test/src/test/java/hudson/model/FreestyleJobPublisherTest.java:
```diff
@@ -0,0 +1,89 @@
+        p.getPublishersList().add(new ArtifactArchiver("result.txt", "", false));
+
+        FreeStyleBuild b = p.scheduleBuild2(0).get();
+        assertEquals("Build must fail, because we used FalsePublisher", b.getResult(), Result.FAILURE);
```
`FalsePublisher` -> `IOExceptionPublisher`

KostyaSha replied:
Oh, thanks, was focused on code and not text messages.

orrc commented:
FWIW, the `testFreestyleWithFalsePublisher` test case fails on my machine.

KostyaSha commented:
@orrc interesting, maybe something changed in core. I will try re-trigger PR-build

kohsuke commented:
:+1:

jglick commented on core/src/main/java/hudson/model/AbstractBuild.java:
```diff
@@ -733,8 +734,10 @@ protected final boolean performAllBuildSteps(BuildListener listener, Iterable<?
             String msg = "Publisher " + bs.getClass().getName() + " aborted due to exception";
-            e.printStackTrace(listener.error(msg));
-            LOGGER.log(WARNING, msg, e);
+            if (!(e instanceof AbortException)){
```
If it _is_ an `AbortException`, is the message being printed somewhere else already? Anyway

> aborted due to exception

does not sound right in this context.

kohsuke replied:
The equivalent pattern in the past (of returning `false` from the `perform()` method) left no message, so the patch is consistent with that.

KostyaSha replied:
Looks also weird for me. Will double-check

jglick replied:
> The equivalent pattern in the past (of returning `false` from the `perform()` method) left no message

Sure, but then the plugin author was _deciding_ to leave no message (or more likely to print one to a `BuildListener` earlier). If they _decided_ to throw `AbortException`, they supplied a message, so we should use it.

jglick commented on core/src/main/java/hudson/model/AbstractBuild.java:
```diff
@@ -733,8 +734,10 @@ protected final boolean performAllBuildSteps(BuildListener listener, Iterable<?
 
         private void reportError(BuildStep bs, Throwable e, BuildListener listener, boolean phase) {
             String msg = "Publisher " + bs.getClass().getName() + " aborted due to exception";
```
Need not be computed if we got an `AbortException`, right?

KostyaSha replied:
It can be even more user friendly

daniel-beck commented:
:+1:

jglick commented on test/src/test/java/hudson/model/FreestyleJobPublisherTest.java:
```diff
@@ -0,0 +1,89 @@
+        assertEquals("Build must fail, because we used FalsePublisher", b.getResult(), Result.FAILURE);
+        File file = new File(b.getArtifactsDir(), "result.txt");
+        assertTrue("ArtifactArchiver is executed even prior publisher fails.", file.exists());
+        assertTrue("Second publisher must see FAILURE status", FileUtils.readFileToString(file).equals(Result.FAILURE.toString()));
```
And I would expect to see the detail message from the `AbortException` in the log.

jglick replied:
Should also be able to

```java
j.assertBuildStatusNotContains("\tat ", b);
```

to prove that we suppressed the stack trace.

jglick commented on test/src/test/java/hudson/model/FreestyleJobPublisherTest.java:
```diff
@@ -0,0 +1,89 @@
+    /**
+     * Execute even one of publishers return false. Shows JENKINS-26964 bug.
+     */
+    @Issue("JENKINS-26964")
```
You might want to move `@Issue` up to be a class annotation; or else copy it on the other test cases, assuming they are also manifestations of the bug (or controls).

jglick commented on test/src/test/java/hudson/model/utils/AbortExceptionPublisher.java:
```diff
@@ -0,0 +1,35 @@
+public class AbortExceptionPublisher extends Recorder {
+    @Override
+    public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListener listener) throws InterruptedException, IOException {
+        throw new AbortException("Throwed AbortException from publisher!");
```
Trivial, but the word you want is `threw`.

jglick commented on test/src/test/java/hudson/model/utils/AbortExceptionPublisher.java:
```diff
@@ -0,0 +1,35 @@
+/**
+ * Publisher that throws AbortException
+ */
+public class AbortExceptionPublisher extends Recorder {
```
I notice these are all `Recorder`s, none `Notifier`s which run in a later build phase and which are not expected to affect the `result`. Not sure if it matters for purposes of this PR, just something to consider.

KostyaSha replied:
It is interesting question because i was unable follow all `phase` and `canContinue` from `perform` method logic. `BuildStepMonitor` seems also plays in this party. 
I can wrote more test but i have no idea how all this statuses should work. AbortException was just added and seems there is no any existed tests for this statuses.

kohsuke commented:
Tests do pass for me.

jglick commented on test/src/test/java/hudson/model/utils/ResultWriterPublisher.java:
```diff
@@ -0,0 +1,47 @@
+    @Override
+    public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListener listener) throws InterruptedException, IOException {
+        FilePath file = build.getWorkspace().child(fileName);
+        file.write(build.getResult().toString(), Charset.defaultCharset().name());
```
This is fine, though it would be slightly more concise to just print the result in a specially-formatted message to the `listener`. Then you can use `JenkinsRule.assertBuildLogContains`.

jglick replied:
…though perhaps not, since you seem to be assuming that a following `ArtifactArchiver` will see it, too.

KostyaSha replied:
Interaction with files is not ideal, but it works very well. buildlog can be screwed or has something wrong. File is strict.

jglick commented on test/src/test/java/hudson/model/FreestyleJobPublisherTest.java:
```diff
@@ -0,0 +1,89 @@
+        assertEquals("Build must fail, because we used FalsePublisher", b.getResult(), Result.FAILURE);
+        File file = new File(b.getArtifactsDir(), "result.txt");
+        assertTrue("ArtifactArchiver is executed even prior publisher fails.", file.exists());
+        assertTrue("Second publisher must see FAILURE status", FileUtils.readFileToString(file).equals(Result.FAILURE.toString()));
```
Actually this was the third publisher (`ResultWriterPublisher`) we are talking about, right?

jglick commented on test/src/test/java/hudson/model/FreestyleJobPublisherTest.java:
```diff
@@ -0,0 +1,89 @@
+        p.getPublishersList().add(new ArtifactArchiver("result.txt", "", false));
+
+        FreeStyleBuild b = p.scheduleBuild2(0).get();
+        assertEquals("Build must fail, because we used FalsePublisher", b.getResult(), Result.FAILURE);
```
BTW these two lines could be collapsed to

```java
FreeStyleBuild b = j.assertBuildStatus(Result.FAILURE, p.scheduleBuild2(0));
```

if you care.

KostyaSha replied:
AFAIR i was not satisfied with it's implementation

jglick commented on test/src/test/java/hudson/model/FreestyleJobPublisherTest.java:
```diff
@@ -0,0 +1,89 @@
+        FreeStyleBuild b = p.scheduleBuild2(0).get();
+        assertEquals("Build must fail, because we used FalsePublisher", b.getResult(), Result.FAILURE);
+        File file = new File(b.getArtifactsDir(), "result.txt");
+        assertTrue("ArtifactArchiver is executed even prior publisher fails.", file.exists());
```
…because `ArtifactArchiver.onlyIfSuccessful` defaults to false, which is not apparent from looking at the test source. You could actually collapse `ResultWriterPublisher` and `ArtifactArchiver` into one ``ArtifactArchiver` with `onlyIfSuccessful` set to true to assert that an artifact is _not_ archived in case the previous publisher fails, which is I guess the key point for this PR.

jglick replied:
(Not *requesting* that you do so, just an idea if you want to pick it up. It would be polite to add a comment explaining that `ArtifactArchiver` by default runs even on already-failed builds.)

KostyaSha replied:
Having two publishers is better, first writes internal build result status that this PR is about, only by dumping from publisher this result state i can check what publisher actually see. 
ArtifactArchiver was used from some other tests, just for transferring dumped file to build dir. I changed code to use `setOnlyIfSuccessful(false)`. Writing to buildLog some `hash status hash` and check for it may simplify test, but i have no strong opinion about it... Also just to be very save another one publisher that will be not executed will fail test.

jglick commented on test/src/test/java/hudson/model/FreestyleJobPublisherTest.java:
```diff
@@ -0,0 +1,89 @@
+        assertEquals("Build must fail, because we used FalsePublisher", b.getResult(), Result.FAILURE);
+        File file = new File(b.getArtifactsDir(), "result.txt");
+        assertTrue("ArtifactArchiver is executed even prior publisher fails.", file.exists());
+        assertTrue("Second publisher must see FAILURE status", FileUtils.readFileToString(file).equals(Result.FAILURE.toString()));
```
```java
j.assertBuildStatusContains("\tat hudson.model.utils.IOExceptionPublisher", b);
```

to prove that we still get the stack trace.

jglick commented:
Reviewed. I would be :+1: if `AbortException.message` were printed. All else is minor, mostly suggestions for strengthening the tests.

KostyaSha commented:
Thanks for super review :) I will fix comments

KostyaSha committed:
Fix PR comments

- Print publisher display name instead of class, so user can understand what publisher in UI was used and failed.
- Setter for artifact archiver shows how it handle status
- Minor typos
- Added test for stacktraces in build log

KostyaSha commented on core/src/main/java/hudson/model/AbstractBuild.java:
```diff
@@ -723,21 +725,33 @@ protected final boolean performAllBuildSteps(BuildListener listener, Iterable<?
+                        r = false;
                     } catch (LinkageError e) {
                         reportError(bs, e, listener, phase);
+                        r = false;
```
`performAllBuildSteps` is ignored in most calls, but i think this should add consistency for failed execution

KostyaSha committed:
Note JENKINS-26964

KostyaSha commented:
Also rebased and noted :sweat_smile: 
I can wrote more tests but i have no ideas what else can be tested.

jglick commented on core/src/main/java/hudson/model/AbstractBuild.java:
```diff
@@ -723,21 +725,33 @@ protected final boolean performAllBuildSteps(BuildListener listener, Iterable<?
                     try {
                         if (!perform(bs,listener)) {
                             LOGGER.log(Level.FINE, "{0} : {1} failed", new Object[] {AbstractBuild.this, bs});
+                            setResult(Result.FAILURE);
```
Does this not need to be conditional on `phase`, like it is in `reportError`?

KostyaSha replied:
Interesting question.

KostyaSha replied:
Changed to have consistency.

jglick commented on core/src/main/java/hudson/model/AbstractBuild.java:
```diff
@@ -723,21 +725,33 @@ protected final boolean performAllBuildSteps(BuildListener listener, Iterable<?
+            if (e instanceof AbortException) {
+                LOGGER.log(Level.FINE, "{0} : {1} failed", new Object[] {AbstractBuild.this, publisher});
+                listener.error("Publisher '" + publisher + "' failed: ");
+                listener.error(e.getMessage());
```
Could make these a single line. I see no reason to print two lines in the log for one failure.

KostyaSha replied:
+

jglick commented on core/src/main/java/hudson/model/AbstractBuild.java:
```diff
@@ -723,21 +725,33 @@ protected final boolean performAllBuildSteps(BuildListener listener, Iterable<?
+            final String publisher = ((Publisher) bs).getDescriptor().getDisplayName();
+
+            if (e instanceof AbortException) {
+                LOGGER.log(Level.FINE, "{0} : {1} failed", new Object[] {AbstractBuild.this, publisher});
```
If you are going to bother logging something at `FINE`, you might as well include the stack trace too.

```java
LOGGER.log(Level.FINE, AbstractBuild.this + " failed in " + bs, e);
```

KostyaSha replied:
AbortException is expected error, you shouldn't print stacktrace or class name. Just publisher name and it's error message.

jglick commented on test/src/test/java/hudson/model/FreestyleJobPublisherTest.java:
```diff
@@ -0,0 +1,106 @@
+package hudson.model;
```
license header

KostyaSha replied:
Why do you really need it? Default license is MIT and i marked as `@author` for test class?

jglick commented on test/src/test/java/hudson/model/FreestyleJobPublisherTest.java:
```diff
@@ -0,0 +1,106 @@
+        FreeStyleBuild b = p.scheduleBuild2(0).get();
+
+        assertEquals("Build must fail, because we used AbortExceptionPublisher", b.getResult(), Result.FAILURE);
+        j.assertLogNotContains("\tat", b); // log must not contain stacktrace
```
Still recommend checking that the `AbortException.message` appears in the log.

KostyaSha replied:
+

jglick commented on test/src/test/java/hudson/model/FreestyleJobPublisherTest.java:
```diff
@@ -0,0 +1,106 @@
+        j.assertLogNotContains("\tat", b); // log must not contain stacktrace
+        File file = new File(b.getArtifactsDir(), "result.txt");
+        assertTrue("ArtifactArchiver is executed even prior publisher fails", file.exists());
+        assertTrue("Second publisher must see FAILURE status",
```
Third publisher, right?

KostyaSha replied:
:angry:

jglick commented on test/src/test/java/hudson/model/utils/IOExceptionPublisher.java:
```diff
@@ -0,0 +1,35 @@
+public class IOExceptionPublisher extends Recorder {
+    @Override
+    public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListener listener) throws InterruptedException, IOException {
+        throw new IOException("Throwed IOException from publisher!");
```
(again, `threw`)

KostyaSha committed:
Fixing PR comments:

- Typo
- Print error combined
- Set result only when phase like in other code parts.
- Test for expected error message in log

jglick commented:
The last PR build failed randomly (timeout after the notorious Credentials-related errors I can never track down), so [submitted for validated merge](https://jenkins.ci.cloudbees.com/job/core/job/jenkins-core-validated-merge/362/) rather than merging immediately.

jglick commented on core/src/main/java/hudson/model/AbstractBuild.java:
```diff
@@ -724,20 +726,33 @@ protected final boolean performAllBuildSteps(BuildListener listener, Iterable<?
-            String msg = "Publisher " + bs.getClass().getName() + " aborted due to exception";
-            e.printStackTrace(listener.error(msg));
-            LOGGER.log(WARNING, msg, e);
+            final String publisher = ((Publisher) bs).getDescriptor().getDisplayName();
```
#1851

KostyaSha replied:
Note, IDE or static analysing tools didn't highlight this code.
